waku/waku_lightpush/protocol.nim

```diff
@@ -26,92 +26,90 @@ type WakuLightPush* = ref object of LPProtocol
   peerManager*: PeerManager
   pushHandler*: PushMessageHandler
   requestRateLimiter*: RequestRateLimiter
-  sharding: Sharding
+  autoSharding: Option[Sharding]
+
+proc handleRequest(
+    wl: WakuLightPush, peerId: PeerId, pushRequest: LightpushRequest
+): Future[WakuLightPushResult] {.async.} =
+  let pubsubTopic = pushRequest.pubSubTopic.valueOr:
+    if wl.autoSharding.isNone():
+      let msg = "Pubsub topic must be specified when static sharding is enabled"
+      error "lightpush request handling error", error = msg
+      return WakuLightPushResult.err(
+        (code: LightpushStatusCode.INVALID_MESSAGE_ERROR, desc: some(msg))
+      )
+
+    let parsedTopic = NsContentTopic.parse(pushRequest.message.contentTopic).valueOr:
+      let msg = "Invalid content-topic:" & $error
+      error "lightpush request handling error", error = msg
+      return WakuLightPushResult.err(
+        (code: LightpushStatusCode.INVALID_MESSAGE_ERROR, desc: some(msg))
+      )
+
+    wl.autoSharding.get().getShard(parsedTopic).valueOr:
+      let msg = "Autosharding error: " & error
+      error "lightpush request handling error", error = msg
+      return WakuLightPushResult.err(
+        (code: LightpushStatusCode.INTERNAL_SERVER_ERROR, desc: some(msg))
+      )
+
+  # ensure checking topic will not cause error at gossipsub level
+  if pubsubTopic.isEmptyOrWhitespace():
+    let msg = "topic must not be empty"
+    error "lightpush request handling error", error = msg
+    return
+      WakuLightPushResult.err((code: LightpushStatusCode.BAD_REQUEST, desc: some(msg)))
+
+  waku_lightpush_v3_messages.inc(labelValues = ["PushRequest"])
+
+  let msg_hash = pubsubTopic.computeMessageHash(pushRequest.message).to0xHex()
+  notice "handling lightpush request",
+    my_peer_id = wl.peerManager.switch.peerInfo.peerId,
+    peer_id = peerId,
+    requestId = pushRequest.requestId,
+    pubsubTopic = pushRequest.pubsubTopic,
+    msg_hash = msg_hash,
+    receivedTime = getNowInNanosecondTime()
+
+  let handleRes = await wl.pushHandler(peerId, pubsubTopic, pushRequest.message)
+
+  if handleRes.isOk():
+    return ok(handleRes.get())
+
+  return err((code: handleRes.error.code, desc: handleRes.error.desc))
 
 proc handleRequest*(
     wl: WakuLightPush, peerId: PeerId, buffer: seq[byte]
 ): Future[LightPushResponse] {.async.} =
-  let reqDecodeRes = LightpushRequest.decode(buffer)
-  var isSuccess = false
   var pushResponse: LightpushResponse
```

Comment:
Storing results in a `var` is usually a bad smell because bigger nesting is needed (instead of just exiting on an error).
Bigger nesting = less readable code.
Proof-in-pudding: some `return` were including in nesting leading to metrics being missed: https://github.com/waku-org/nwaku/pull/3468/files#r2168086052